File: M/2130-maximum-twin-sum-of-a-linked-list/main.go
package main

type ListNode struct {
	Val  int
	Next *ListNode
}

func pairSum(head *ListNode) int {
	cnt := 0

	p := head
	for p != nil {
		p = p.Next
		cnt++
	}
	buf := make([]int, cnt/2)
	s := 0
	p = head
	for s < cnt/2 {
		buf[s] = p.Val
		s++
		p = p.Next
	}

	s--
	for s >= 0 {
		buf[s] += p.Val
		p = p.Next
		s--
	}
	max := buf[0]
	for i := 1; i < cnt/2; i++ {
		if max < buf[i] {
			max = buf[i]
		}
	}
	return max
}

File: M/2130-maximum-twin-sum-of-a-linked-list/main.rs
impl Solution {
  pub fn pair_sum(head: Option<Box<ListNode>>) -> i32 {
    let mut buf: Vec<i32> = vec![];
    let mut h = head;
    while let Some(ln) = h {
      buf.push(ln.val);
      h = ln.next;
    }

    (0..buf.len() / 2).fold(i32::MIN, |ans, idx| ans.max(buf[idx] + buf[buf.len() - 1 - idx]))
  }
}

// Definition for singly-linked list.
// #[derive(PartialEq, Eq, Clone, Debug)]
// pub struct ListNode {
//   pub val: i32,
//   pub next: Option<Box<ListNode>>
// }
//
// impl ListNode {
//   #[inline]
//   fn new(val: i32) -> Self {
//     ListNode {
//       next: None,
//       val
//     }
//   }
// }
